### PR TITLE
Bug 1943378: Eliminate instanceCreate volume leak

### PR DESCRIPTION
**Pull request by pierreprinetti**
InstanceCreate is a method that creates a server. If the server is set
to boot from volume, and an image ID is passed, the method creates the
root volume prior to creating the server. The root volume is set to be
destroyed when the associated server is destroyed.

However, if the server fails to create (for example, because of quota
issues), the volume is never associated to a server and the automatic
deletion is never triggered. At every round of retry, a new volume will
be created, possibly until volume quota is reached (or server creation
is successful). This results in a leakage of unused volumes.

With this patch, a newly created root volume is explicitly deleted as
soon as the server creation call fails.

Note that this patch leaves unmodified the lifespan of a volume
associated to a server, regardless if the server ever reaches an ACTIVE
state.

Co-Authored-By: Matthew Booth <mbooth@redhat.com>

### CODE REVIEW

**Review by mandre — COMMENTED** [review]
Seems like an easy enough fix. Have you checked if k8s/capo also exhibits the problem and thus needs a similar fix?

**Comment by pierreprinetti** [comment]
@mdbooth 

I figured that the "quota exceeded" case fails synchronously, and not asynchronously (i.e. `InstanceCreate` returns an error to `Create`, thus the wait-for-ACTIVE loop is never hit). Which makes sense: conversely, a machine that is successfully created likely gets its volume associated, and destroyed upon deletion, even when it never reaches the ACTIVE state.

I tested this patch with local-capo.sh and using MOC's huge flavor. It works: the volume is correctly cleaned up when machine creation fails.

@mdbooth @mandre 

I am aware that the `InstanceCreate` code does not comply with the idempotency criterion for CAPO operations. The volume-deletion request is only kept in memory and won't survive a restart: if CAPO is rebooted while creating a server, the volume might leak.
The reason why I am not tackling that issue, is that I believe it is not a regression. The same would happen before this patch. I have [filed a bug](https://bugzilla.redhat.com/show_bug.cgi?id=1983612) for this; we can have a conversation in BZ on whether it's worth fixing.

**Comment by pierreprinetti** [comment]
/hold cancel

Unit tests are now passing; this patch is ready for a review.

**Comment by pierreprinetti** [comment]
@mandre 

>Have you checked if k8s/capo also exhibits the problem

It does; @mdbooth has done that investigation. Since the bug is high-severity, we fix here then there.

**Comment by mandre** [comment]
Alternatively, instead of keeping track of volume that was created and eventually delete it on failure, we may adopt a different strategy where we check if a volume exists before creating a new one, if it's possible to identify such volume. 

However, if the leaked volume gets deleted on cluster destroy, I believe it's not such a big deal: MAO restarting should be an exceptional event.

Would like to get @mdbooth's opinion on the PR.

**Comment by pierreprinetti** [comment]
@mandre 
> check if a volume exists before creating a new one

This is exactly what occurred to me while thinking about a solution for [the idempotency bug](https://bugzilla.redhat.com/show_bug.cgi?id=1983612) :smile:

File: pkg/cloud/openstack/clients/machineservice.go [hunk 1]
```diff
@@ -730,6 +732,9 @@ func (is *InstanceService) InstanceCreate(clusterName string, name string, clust
 				return nil, fmt.Errorf("Create bootable volume err: %v", err)
 			}
 
+			cleanupOperationsInCaseOfServerCreationFailure = append(cleanupOperationsInCaseOfServerCreationFailure, func() error {
+				return volumes.Delete(is.volumeClient, volume.ID, nil).ExtractErr()
```
Comment (mdbooth) [inline pkg/cloud/openstack/clients/machineservice.go:736]: This differs from the delete line below:

```go
err = volumes.Delete(is.volumeClient, volumeID, volumes.DeleteOpts{}).ExtractErr()
```

Have you managed to test it manually?

Comment (mdbooth) [reply pkg/cloud/openstack/clients/machineservice.go:736]: There's also at least 1 other error condition on line 777 which, if it failed, would result in the same leak.

There might be a slightly more robust solution. How about something like:

```go
defer func() error {
    if server == nil {
        err = volumes.Delete(is.volumeClient, volumeID, volumes.DeleteOpts{}).ExtractErr()
        ... log any error
    }
}
```

**Review by mdbooth — CHANGES_REQUESTED** [review]

**Comment by mdbooth** [comment]
Ah, crap, it's already merged :( We still need to fix it.

File: pkg/cloud/openstack/clients/machineservice.go (hunk 1 above)
Comment (pierreprinetti) [reply pkg/cloud/openstack/clients/machineservice.go:736]: I tested manually and it works. However the `defer` solution is obviously better, I'm back to the keyboard.

Comment (pierreprinetti) [reply pkg/cloud/openstack/clients/machineservice.go:736]: By the way, as far as I can tell, passing an empty `volumes.DeleteOpts{}` is a fancy noop with [lots of reflection](https://github.com/gophercloud/gophercloud/blob/2509091608e7b377980b0d24a3b60e3cc82f251f/params.go#L324) in the way.

**Pull request merged**